### DISCUSSION TITLE
STUB/MOCK qeury result

### DISCUSSION THREAD

**Discussion by MS-24** [utPLSQL/utPLSQL#1331] [Q&A]
Hi ,

Is there a way to stub the result of a query instead actually querying from Database?
eg:

**_actual_code :-_** 

1. procedure actual_code as
2.     begin
3.          --- Some logic 
4.         commit;
5.         select value into l_value from tab_test where name='John';
6.     exception when no_data_found then
7.      raise_appliaction_error(-20400,'Blah blah');
8.    end actual_code;

----------------------------------------------
_**test suite spec**_
--%test (test for excpetion)
--%throw(-20400)
procedure test_exception;

**_test suite body_**

1. procedure test_exception as 
2.    execute immediate 'delete from tab_test where name='''John'''';
3.    actual_code;
4. end test_exception;

-------------------------------------------------------------

I got the result intended with the above setup , but due to commit in line 3 of actual_code , the data gets deleted from the table which is not something wanted .

Could you please suggest if any workaround.

I would like to stub the result as stub_query('tab_test',row[value=>'Student',name=>'John']  or some null, so that it won't refer to actual table in data base. 
Is this something achievable ?

Thanks in advance.

**Comment by lwasylow** [comment]
Wrap the actual code in pragma autonomous procedure as part of test and call it.

**Comment by pesse** [comment]
From my point of view, the main problem might be the design of your `actual_code` procedure. 
Yes, you *can* probably "stub" a query with a sys_refcursor, but what you're most likely experiencing is that you discovered code which is hard to test and therefore maybe also hard to maintain.

You might want to think about separating transactional (commit) and business logic (what should be done) or make your methods only do one thing (like: do_some_logic_with_commit, get_remaining_information).
Situations like this are one of the core benefits of automated self-testing: If something is hard to test, it can probably be implemented in a way which is easier to understand, maintain and test.

GeePaw Hill describes this as "awkward" vs. "graceful" test situations: 
http://geepawhill.org/pro-tip-awkward-and-graceful/

**Comment by MS-24** [comment]
Thanks @lwasylow  and @pesse . 
@pesse this is the one of scenario which involve transactional and business logic together .  There could be other different situations.
I am also thinking it would be faster to execute the test  if we don't have read the actual table(big table) instead the test frame work invokes the stub data when it sees a query on the actual table.
Not very sure , but it is just an idea . 
thanks for the response anyway. Cheers!!
